### ISSUE TITLE
Implement rational numbers exercise

### ISSUE THREAD

**Issue by robkeim** [exercism/csharp#517]
If you want to implement this exercise, please note that you need to ensure the test cases are generated based on the input.

Here's the exercise description:
https://github.com/exercism/problem-specifications/blob/master/exercises/rational-numbers/description.md

Here's the test case data:
https://github.com/exercism/problem-specifications/blob/master/exercises/rational-numbers/canonical-data.json

**Comment by ShamilS** [comment]
@robkeim If nobody else would take it in the coming few days, then I will take it to make it completed within, say 10 days. I have got the tests already generated here [Rational-Numbers (first Try)](https://github.com/ShamilS/csharp/tree/rational-numbers-first-try)

**Comment by ShamilS** [comment]
@ErikSchierboom OK, let's set deadline for the first PR for the 14th of April.

**Comment by robkeim** [comment]
@ShamilS yes please feel free to take this, your help is much appreciated!

**Comment by ShamilS** [comment]
@ErikSchierboom and @robkeim I have actually made the first version, which runs all the tests "green". But there are two issues:

1. Math.Power is used in the following method:

```
    private static double pow2(int baseNumber, int[] power)
    {
        if (power[0] > 0 && power[1] > 0)
        {
            var tempResult = pow(baseNumber, power[0]);
            return Math.Pow((double)tempResult, 1 / (double)power[1]);
        }
        else return Math.Pow((double)baseNumber, (double)power[0] / (double)power[1]);
    }
```

Any hints how to get out of using Math.Power at all?

2. Assert.Equal do not get properly verified to True when they compare equal(?) double numbers - to make the tests pass I've currently had to convert an expected value and a calculated value .ToString().

```
    [Fact()]
    public void Raise_a_real_number_to_a_positive_rational_number()
    {
        //Assert.Equal((double)16, RationalNumbers.Expreal(8, new[] { 4, 3 }));
        Assert.Equal(((double)16).ToString(), RationalNumbers.Expreal(8, new[] { 4, 3 }).ToString());
    }

    [Fact()]
    public void Raise_a_real_number_to_a_negative_rational_number()
    {
        //Assert.Equal((double)0.333333333333333, RationalNumbers.Expreal(9, new[] { -1, 2 }));
        Assert.Equal(((double)0.333333333333333).ToString(), RationalNumbers.Expreal(9, new[] { -1, 2 }).ToString());
    }
```

Any hints how to solve this issue?

![rational-numbers](https://user-images.githubusercontent.com/2162601/38330904-7a095bcc-385a-11e8-80d3-ec04f73375db.png)

**Comment by robkeim** [comment]
@ShamilS regarding the second issue we had the same problem with the `ComplexNumbers` exercise (this is a common problem when dealing with high precision numbers) and the common convention is to check the numbers are equal with a difference threshold. You can see what we did for `ComplexNumbers` here:
https://github.com/exercism/csharp/blob/378ef262f888f6ebed411532fe9fcbbfea835601/generators/Exercises/ComplexNumbers.cs#L57

**Comment by robkeim** [comment]
Here's the MSDN for the `Assert.Equal` overload being used:
https://msdn.microsoft.com/en-us/library/ms243458.aspx

**Comment by robkeim** [comment]
Clicked the wrong button didn't mean to close this issue.

**Comment by ShamilS** [comment]
@robkeim Yes, if the power is an integer it can be simulated by multiplying but the issue is with the cases as:
 
    8 ^ (4/3)  and 9 ^ (-1/2)

For this case logarithms can be probably used? - If logarithms will work would using Math.Log(...) be acceptable for this exercise solution?

**Comment by ErikSchierboom** [comment]
Maybe take a look at the corresponding F# PR: https://github.com/exercism/fsharp/pull/514

**Comment by ShamilS** [comment]
@ErikSchierboom  Yes, **nthroot** method from F# PR should help. I will take my time now, 10 days as planned - should be enough to solve the two issues I have mentioned above + clean the code to make it ready for PR.

**Comment by ShamilS** [comment]
@ErikSchierboom and @robkeim I have just released a PR. Please check.

**Comment by ErikSchierboom** [comment]
Closed by #533.
